executor/adapter.go

```diff
@@ -271,6 +284,102 @@ func (a *ExecStmt) Exec(ctx context.Context) (sqlexec.RecordSet, error) {
 	}, nil
 }
 
+type chunkRowRecordSet struct {
+	rows   []chunk.Row
+	idx    int
+	fields []*ast.ResultField
+	e      Executor
+}
+
+func (c *chunkRowRecordSet) Fields() []*ast.ResultField {
+	return c.fields
+}
+
+func (c *chunkRowRecordSet) Next(ctx context.Context, req *chunk.RecordBatch) error {
+	chk := req.Chunk
+	chk.Reset()
+	for !chk.IsFull() && c.idx < len(c.rows) {
+		chk.AppendRow(c.rows[c.idx])
+		c.idx++
+	}
+	return nil
+}
+
+func (c *chunkRowRecordSet) NewRecordBatch() *chunk.RecordBatch {
+	return chunk.NewRecordBatch(c.e.newFirstChunk())
+}
+
+func (c *chunkRowRecordSet) Close() error {
+	return nil
+}
+
+func (a *ExecStmt) handlePessimisticSelectForUpdate(ctx context.Context, sctx sessionctx.Context, e Executor) (sqlexec.RecordSet, error) {
+	txnCtx := sctx.GetSessionVars().TxnCtx
+	retryCnt := uint(0)
+	for {
+		rs, err := a.runPessimisticSelectForUpdate(ctx, sctx, e)
+		if err == nil {
+			return rs, nil
+		}
+		// Retry this "select for update" statement using a new startTS.
+		if !strings.Contains(err.Error(), tidbutil.WriteConflictMarker) {
```

Comment:
It is better to retry for `try again later`.

Reply:
I suppose 'try again later' is handled by tikv-client.